libraries/chain/controller.cpp

```diff
@@ -1063,6 +1063,36 @@ struct controller_impl {
       });
    }
 
+   template <typename F>
+   std::optional<block_header_state_legacy> fetch_bhs_for_legacy_rpc_by_x(F&& f, const std::optional<block_num_type> block_num = std::nullopt) const {
```

Comment:
It feels wrong for logic specific to how an RPC endpoint operates to be in `controller` (instead of the `chain_plugin` leveraging more generic calls against `controller`), but `controller` seems to be in the best position to create something concise with `forkdb` & `apply()`

Reply:
We are trying to keep `block_header_state` and `block_header_state_legacy` from leaking out of controller.  Seems like this should return a `signed_block_ptr` and move the extraction of data into `chain_plugin`. You should then be able to just use the existing `fetch_block_by_id` and `fetch_block_by_number`.